web/src/lib/components/settings/Skills.svelte

```diff
@@ -0,0 +1,117 @@
+<script lang="ts">
+    import { onMount } from 'svelte';
+    import { ErrorHandler, putProfile, getEntities, postEntity, putEntity } from "$lib/services/api";
+    import { Info, Error, token, user } from "$lib/stores";
+    import { page } from '$app/stores';
+    import { UserAchievement, userAchievementFromJson } from "$lib/types/user";
+    import ListView, { ListViewStyle } from "$lib/components/ListView.svelte";
+    import UserAchievementCard from "$lib/components/UserAchievementCard.svelte";
+    import UserAchievementEditor from "$lib/components/UserAchievementEditor.svelte";
+
+    export let onSave: () => void = () => {};
+
+    let skills: string[] = [];
+    let achievements: UserAchievement[] = [];
+
+    let newSkill: string = "";
+
+    $: addedSkills = skills.filter(s => !$user || $user.skills.indexOf(s) === -1);
+    $: removedSkills = $user ? $user.skills.filter(s => skills.indexOf(s) === -1) : [];
+
+    $: saveButtonActive = $user && !saving && (addedSkills.length !== 0 || removedSkills.length !== 0);
+
+    function skillKeyPress(e) {
+        if (e.key === "Enter") {
+            if (newSkill.length > 21) {
+                Error.set("Skill should be less than 21 characters long.");
+                return;
+            }
+            skills = [...skills, newSkill]; // NB: push won't work, as Svelte won't detect the change to the array
+            newSkill = "";
+        }
+    }
+
+    function removeSkill(skill) {
+        let i = skills.indexOf(skill);
+        skills = skills.slice(0, i).concat(skills.slice(i + 1));
```

Comment:
It's not that you need ultra-performance here, but maybe a little simpler code by changing that to `splice`?

Reply:
Actually I did `splice` first, but that doesn't work.

The way Svelte works, it looks for assignment to detect that something has changed and update the UI, otherwise the array would change but Svelte would not do anything. :(

Reply:
Ahh, I understand...

I never used those, but I think that may be a possible use case for that "reactive blocks":

```